tools/src/main/scala/com/nvidia/spark/rapids/tool/profiling/AutoTuning.scala

```diff
@@ -0,0 +1,616 @@
+   */
+  def calcGpuConcTasks(): Long = {
+    // spark.rapids.sql.concurrentGpuTasks
+    Math.min(MAX_CONC_GPU_TASKS,
+      convertToMB(clusterProps.gpu.memory) / DEF_GPU_MEM_PER_TASK_MB)
+  }
+
+  private def calcExecutorContainerMem(): Double = {
+    // account for system overhead
+    val usableWorkerMem =
+      Math.max(0, convertToMB(clusterProps.system.memory) - DEF_SYSTEM_RESERVE_MB)
+    val executorsPerNode = clusterProps.gpu.getCount
+    if (executorsPerNode == 0) {
+      0.0
+    } else {
+      (1.0 * usableWorkerMem) / executorsPerNode
+    }
+  }
+
+  /**
+   * Recommendation for 'spark.executor.memory' based on system memory, cluster scheduler
+   * and num of gpus.
+   */
+  def calcExecutorMemory(executorContainerMemCalculator: () => Double,
+      numExecCoresClaculator: () => Int): Long = {
+    // spark.executor.memory
+    // reserve 10% of heap as memory overhead
+    val maxExecutorHeap = Math.max(0,
+      executorContainerMemCalculator.apply() * (1 - DEF_HEAP_OVERHEAD_FRACTION)).toInt
+    // give up to 2GB of heap to each executor core
+    Math.min(maxExecutorHeap, DEF_HEAP_PER_CORE_MB * numExecCoresClaculator.apply())
+  }
+
+  def calcPinnedMemoryWithOverhead(
+      execMemoryCalculator: (() => Double, () => Int) => Double,
+      containerMemCalculator: () => Double,
+      execCoresClaculator: () => Int): (Long, Long) = {
+    // spark.rapids.memory.pinnedPool.size, spark.executor.memoryOverhead
+    val executorHeap =
+      execMemoryCalculator.apply(containerMemCalculator, execCoresClaculator)
+    var executorMemOverhead = (executorHeap * DEF_HEAP_OVERHEAD_FRACTION).toLong
+    // pinned memory uses any unused space up to 4GB
+    val pinnedMem = Math.min(MAX_PINNED_MEMORY_MB,
+      containerMemCalculator.apply() - executorHeap - executorMemOverhead).toLong
+    executorMemOverhead += pinnedMem
+    (pinnedMem, executorMemOverhead)
+  }
+
+  def calculateRecommendations(): Unit = {
+    appendRecommendation("spark.executor.instances", s"${calcExecInstances()}")
+    // spark.executor.cores
+    appendRecommendation("spark.executor.cores", s"${calcNumExecutorCores()}")
+    // spark.task.resource.gpu.amount
+    appendRecommendation("spark.task.resource.gpu.amount",
+      s"${calcTaskGPUAmount(calcNumExecutorCores)}")
+    // spark.rapids.sql.concurrentGpuTasks
+    appendRecommendation("spark.rapids.sql.concurrentGpuTasks",
+      s"${calcGpuConcTasks()}")
+    appendRecommendationForMemoryMB("spark.executor.memory",
+      s"${calcExecutorMemory(calcExecutorContainerMem, calcNumExecutorCores)}")
+    val (pinnedMemory, memoryOverhead) =
+      calcPinnedMemoryWithOverhead(
+        calcExecutorMemory, calcExecutorContainerMem, calcNumExecutorCores)
+    appendRecommendationForMemoryMB("spark.executor.memoryOverhead", s"$memoryOverhead")
+    appendRecommendationForMemoryMB("spark.rapids.memory.pinnedPool.size", s"$pinnedMemory")
```

Comment:
if pinnedMemory is 0 we shouldn't add the config

Reply:
`appendRecommendation` won't add the element if the value is null. this saves a lot of redundant checks across all recommendations.